### ISSUE TITLE
Support cross-validation for selecting the best weight decay / L2 penalty

### ISSUE THREAD

**Issue by norabelrose** [EleutherAI/elk#21]
When @FabienRoger started playing around with the LBFGS optimizer he had to manually tune the L2 penalty to prevent overfitting on small datasets, which is not ideal. There is a simple well-understood way of setting regularization terms like this: [cross-validation](https://en.wikipedia.org/wiki/Cross-validation_(statistics)). Scikit-learn implements this for logistic regression as [`LogisticRegressionCV`](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegressionCV.html)— which, by the way, we arguably should be using instead of the vanilla `LogisticRegression` class to get a better supervised baseline.

We'd use a logarithmic grid of, say, 10 L2 penalties and use the one which achieves the best _out-of-sample_ unsupervised loss on the held out folds of the training data. After selecting the best penalty you'd do one last fit on the whole training dataset with that penalty.

This is primarily useful for smallish datasets where the number of datapoints is roughly equal to, or even smaller than, the hidden state dimensionality. It turns out, though, that a lot of the datasets we're looking at are of this size, so this does matter. The main downside of doing this is that it slows down the fitting process, but LBFGS is already quite a bit faster than Adam so it may not be much of a slowdown relative to the original code.

I'd say this isn't quite as pressing as some of the other issues right now but it would definitely improve the usability and robustness of the method. Since I've implemented cross validation methods before I'd be happy to do this myself when I have time but if someone else wants to take it on that'd be great.

**Comment by norabelrose** [comment]
One problem which might make this tricky is that the loss functions we're using are not convex, which means performance depends on the initialization. We may have to take the max performance over N random seeds for each penalty, or something.

We can potentially speed this up substantially by sequentially iterating over the grid of L2 penalties and _warm-starting_ the optimization process for each penalty value with the parameters found for the previous penalty value, [as scikit-learn does](https://scikit-learn.org/stable/glossary.html#term-warm_start) when `warm_start=True`.

**Comment by norabelrose** [comment]
Fixed by #135
